third/nmh/man/rcvstore.man
.\"
.\" %nmhwarning%
.\" $Id: rcvstore.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH RCVSTORE %manext1% MH.6.8 [%nmhversion%]
.SH NAME
rcvstore \- asynchronously incorporate mail into a folder
.SH SYNOPSIS
.in +.5i
.ti -.5i
%libdir%/rcvstore
\%[+folder]
.br
\%[\-create] \%[\-nocreate]
\%[\-unseen] \%[\-nounseen]
.br
\%[\-zero] \%[\-nozero]
\%[\-public] \%[\-nopublic]
.br
\%[\-sequence\ name\ ...]
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fIRcvstore\fR incorporates a message from the standard input into an
\fInmh\fR folder.  This command is typically used in conjunction with
mail filtering programs such as \fIslocal\fR and \fIprocmail\fR, to
filter your mail into different folders.

You may specify which folder to use with `+folder'.  If no folder is
specified, \fIrcvstore\fP will use the folder given by a non\-empty
\*(lqInbox:\*(rq entry in the user's profile, else it will use the folder
named \*(lqinbox\*(rq.

If the switch `\-create' is given (it is the default) and if the specified
(or default) folder does not exist, then it will be created.  You may
disable this with the `\-nocreate' option.  In this case \fIrcvstore\fP
will exit if the specified folder does not exist.

When the new message is incorporated into the folder, it is assigned
the next highest number for that folder.

\fIRcvstore\fR will incorporate anything except zero length messages
into the user's nmh folder.  It will not change the message in any
way.

If the user's profile contains a \*(lqMsg\-Protect: nnn\*(rq entry, it
will be used as the protection on the newly created message, otherwise
the \fInmh\fR default of 0644 will be used.  For all subsequent operations
on this message, this initially assigned protection will be preserved.

If the switch `\-unseen' is given (it is on by default), and if the
profile entry \*(lqUnseen\-Sequence\*(rq is present and non\-empty, then
\fIrcvstore\fR will add the newly incorporated message to each sequence
named by this profile entry.  You may use the switch `\-nounseen' to
disable this.  These sequences will not be zero'ed by \fIrcvstore\fR
prior to adding the new message.

Furthermore, the incoming message may be added to additional sequences
as they arrive by the use of the `\-sequence' switch.  As with the
commands \fIpick\fP and \fImark\fP, you may also use the switches
`\-zero' and `\-nozero' to specify whether to zero old sequences or not.
Similarly, use of the `\-public' and `\-nopublic switches may be used
to force these sequences to be public or private sequences.

.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Folder\-Protect:~^To set mode when creating a new folder
.Ps
^Inbox:~^To find the default inbox
.Ps
^Msg\-Protect:~^To set mode when creating a new message
.Ps
^Unseen\-Sequence:~^To name sequences denoting unseen messages
.Sa
rcvdist(1), rcvpack(1), rcvtty(1), mh\-sequence(5)
.De
`+folder' defaults to \*(lqInbox\*(rq profile entry
.Ds
`\-create'
.Ds
`\-unseen'
.Ds
`\-nozero'
.Co
No context changes will be attempted, with the exception of
sequence manipulation.
.Bu
If you use the \*(lqUnseen\-Sequence\*(rq profile entry, \fIrcvstore\fP
could try to update the context while another \fInmh\fP process
is also trying to do so.  This can cause the context to become
corrupted.  To avoid this, do not use \fIrcvstore\fP if you use the
\*(lqUnseen\-Sequence\*(rq profile entry.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/rcvstore.c

/*
 * rcvstore.c -- asynchronously add mail to a folder
 *
 * $Id: rcvstore.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>
#include <h/signals.h>
#include <errno.h>
#include <signal.h>

static struct swit switches[] = {
#define CRETSW         0
    { "create",	0 },
#define NCRETSW        1
    { "nocreate", 0 },
#define UNSEENSW       2
    { "unseen", 0 },
#define NUNSEENSW      3
    { "nounseen", 0 },
#define PUBSW          4
    { "public",	0 },
#define NPUBSW         5
    { "nopublic",  0 },
#define ZEROSW         6
    { "zero",	0 },
#define NZEROSW        7
    { "nozero",	0 },
#define SEQSW          8
    { "sequence name", 0 },
#define VERSIONSW      9
    { "version", 0 },
#define HELPSW        10
    { "help", 4 },
    { NULL, 0 }
};

extern int errno;

/*
 * name of temporary file to store incoming message
 */
static char *tmpfilenam = NULL;


int
main (int argc, char **argv)
{
    int publicsw = -1, zerosw = 0;
    int create = 1, unseensw = 1;
    int fd, msgnum, seqp = 0;
    char *cp, *maildir, *folder = NULL, buf[BUFSIZ];
    char **argp, **arguments, *seqs[NUMATTRS+1];
    struct msgs *mp;
    struct stat st;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    mts_init (invo_name);
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    /* parse arguments */
    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
	    case AMBIGSW: 
		ambigsw (cp, switches);
		done (1);
	    case UNKWNSW: 
		adios (NULL, "-%s unknown", cp);

	    case HELPSW: 
		snprintf (buf, sizeof(buf), "%s [+folder] [switches]",
			  invo_name);
		print_help (buf, switches, 1);
		done (1);
	    case VERSIONSW:
		print_version(invo_name);
		done (1);

	    case SEQSW: 
		if (!(cp = *argp++) || *cp == '-')
		    adios (NULL, "missing argument name to %s", argp[-2]);

		/* check if too many sequences specified */
		if (seqp >= NUMATTRS)
		    adios (NULL, "too many sequences (more than %d) specified", NUMATTRS);
		seqs[seqp++] = cp;
		continue;

	    case UNSEENSW:
		unseensw = 1;
		continue;
	    case NUNSEENSW:
		unseensw = 0;
		continue;

	    case PUBSW: 
		publicsw = 1;
		continue;
	    case NPUBSW: 
		publicsw = 0;
		continue;

	    case ZEROSW: 
		zerosw++;
		continue;
	    case NZEROSW: 
		zerosw = 0;
		continue;

	    case CRETSW: 
		create++;
		continue;
	    case NCRETSW: 
		create = 0;
		continue;
	    }
	}
	if (*cp == '+' || *cp == '@') {
	    if (folder)
		adios (NULL, "only one folder at a time!");
	    else
		folder = path (cp + 1, *cp == '+' ? TFOLDER : TSUBCWF);
	} else {
	    adios (NULL, "usage: %s [+folder] [switches]", invo_name);
	}
    }

    seqs[seqp] = NULL;	/* NULL terminate list of sequences */

    if (!context_find ("path"))
	free (path ("./", TFOLDER));

    /* if no folder is given, use default folder */
    if (!folder)
	folder = getfolder (0);
    maildir = m_maildir (folder);

    /* check if folder exists */
    if (stat (maildir, &st) == NOTOK) {
	if (errno != ENOENT)
	    adios (maildir, "error on folder");
	if (!create)
	    adios (NULL, "folder %s doesn't exist", maildir);
	if (!makedir (maildir))
	    adios (NULL, "unable to create folder %s", maildir);
    }

    if (chdir (maildir) == NOTOK)
	adios (maildir, "unable to change directory to");

    /* ignore a few signals */
    SIGNAL (SIGHUP, SIG_IGN);
    SIGNAL (SIGINT, SIG_IGN);
    SIGNAL (SIGQUIT, SIG_IGN);
    SIGNAL (SIGTERM, SIG_IGN);

    /* create a temporary file */
    tmpfilenam = m_scratch ("", invo_name);
    if ((fd = creat (tmpfilenam, m_gmprot ())) == NOTOK)
	adios (tmpfilenam, "unable to create");
    chmod (tmpfilenam, m_gmprot());

    /* copy the message from stdin into temp file */
    cpydata (fileno (stdin), fd, "standard input", tmpfilenam);

    if (fstat (fd, &st) == NOTOK) {
	unlink (tmpfilenam);
	adios (tmpfilenam, "unable to fstat");
    }
    if (close (fd) == NOTOK)
	adios (tmpfilenam, "error closing");

    /* don't add file if it is empty */
    if (st.st_size == 0) {
	unlink (tmpfilenam);
	advise (NULL, "empty file");
	done (0);
    }

    /*
     * read folder and create message structure
     */
    if (!(mp = folder_read (folder)))
	adios (NULL, "unable to read folder %s", folder);

    /*
     * Link message into folder, and possibly add
     * to the Unseen-Sequence's.
     */
    if ((msgnum = folder_addmsg (&mp, tmpfilenam, 0, unseensw, 0)) == -1)
	done (1);

    /*
     * Add the message to any extra sequences
     * that have been specified.
     */
    for (seqp = 0; seqs[seqp]; seqp++) {
	if (!seq_addmsg (mp, seqs[seqp], msgnum, publicsw, zerosw))
	    done (1);
    }

    seq_setunseen (mp, 0);	/* synchronize any Unseen-Sequence's      */
    seq_save (mp);		/* synchronize and save message sequences */
    folder_free (mp);		/* free folder/message structure          */

    context_save ();		/* save the global context file           */
    unlink (tmpfilenam);	/* remove temporary file                  */
    tmpfilenam = NULL;

    done (0);
}

/*
 * Clean up and exit
 */
void
done(int status)
{
    if (tmpfilenam && *tmpfilenam)
	unlink (tmpfilenam);
    exit (status);
}
